app/controllers/uploads_controller.rb

```diff
@@ -9,15 +9,44 @@ class UploadsController < ApplicationController
   protect_from_forgery except: :show
 
   before_action :is_asset_path, :apply_cdn_headers, only: [:show, :show_short, :show_secure]
-  before_action :external_store_check, only: [:show_secure, :generate_presigned_put, :complete_external_upload]
+  before_action :external_store_check, only: [
+    :show_secure,
+    :generate_presigned_put,
+    :complete_external_upload,
+    :create_multipart,
+    :batch_presign_multipart_parts,
+    :abort_multipart,
+    :complete_multipart
+  ]
+  before_action :direct_s3_uploads_check, only: [
+    :generate_presigned_put,
+    :complete_external_upload,
+    :create_multipart,
+    :batch_presign_multipart_parts,
+    :abort_multipart,
+    :complete_multipart
+  ]
+  before_action :can_upload_external?, only: [:create_multipart, :generate_presigned_put]
 
   SECURE_REDIRECT_GRACE_SECONDS = 5
-  PRESIGNED_PUT_RATE_LIMIT_PER_MINUTE = 5
+  PRESIGNED_PUT_RATE_LIMIT_PER_MINUTE = 10
+  CREATE_MULTIPART_RATE_LIMIT_PER_MINUTE = 10
+  COMPLETE_MULTIPART_RATE_LIMIT_PER_MINUTE = 10
+  ABORT_MULTIPART_RATE_LIMIT_PER_MINUTE = 10
+  BATCH_PRESIGN_RATE_LIMIT_PER_MINUTE = 10
 
   def external_store_check
     return render_404 if !Discourse.store.external?
   end
 
+  def direct_s3_uploads_check
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+  end
+
+  def can_upload_external?
+    raise Discourse::InvalidAccess if !guardian.can_upload_external?
+  end
```

Comment:
Let's move all these methods to be `private`. Usually the only public methods in a controller are the controller actions.